signature/pipeline_invariants_test.go

```diff
@@ -0,0 +1,179 @@
+	}
+
+	fields := []string{"command", "env", "plugins", "matrix", "repository_url", "secrets"}
+	values, err := step.ValuesForFields(fields)
+	if err != nil {
+		t.Fatalf("ValuesForFields() error = %v", err)
+	}
+
+	// Check that secrets value is included
+	if _, has := values["secrets"]; !has {
+		t.Errorf("ValuesForFields() missing 'secrets' field")
+	}
+
+	// Check that secrets field contains expected data
+	secrets, ok := values["secrets"].(pipeline.Secrets)
+	if !ok {
+		t.Errorf("ValuesForFields() 'secrets' field is not pipeline.Secrets type, got %T", values["secrets"])
+	}
+
+	if len(secrets) != 1 {
+		t.Errorf("ValuesForFields() expected 1 secret, got %d", len(secrets))
+	}
+
+	if secrets[0].Key != "DATABASE_URL" {
+		t.Errorf("ValuesForFields() expected secret key 'DATABASE_URL', got %q", secrets[0].Key)
+	}
+}
+
+func TestCommandStepWithInvariants_ValuesForFields_MissingSecretsField(t *testing.T) {
+	t.Parallel()
+
+	dbUrl := "DATABASE_URL"
+	step := commandStepWithInvariants{
+		CommandStep: pipeline.CommandStep{
+			Command: "echo hello",
+			Env:     map[string]string{"FOO": "bar"},
+			Plugins: pipeline.Plugins{},
+			Secrets: pipeline.Secrets{
+				{Key: "DATABASE_URL", EnvironmentVariable: &dbUrl},
+			},
+		},
+		RepositoryURL: "https://github.com/example/repo",
+		OuterEnv:      map[string]string{"PIPELINE_VAR": "value"},
+	}
+
+	// Request fields without secrets - should fail
+	fields := []string{"command", "env", "plugins", "matrix", "repository_url"}
+	_, err := step.ValuesForFields(fields)
+	if err == nil {
+		t.Fatalf("ValuesForFields() expected error when secrets field not requested but step has secrets")
+	}
+
+	expectedError := "one or more required fields are not present: [secrets]"
+	if err.Error() != expectedError {
+		t.Errorf("ValuesForFields() expected error %q, got %q", expectedError, err.Error())
+	}
+}
+
+func TestCommandStepWithInvariants_ValuesForFields_NoSecretsNoSecretsField(t *testing.T) {
+	t.Parallel()
+
+	step := commandStepWithInvariants{
+		CommandStep: pipeline.CommandStep{
+			Command: "echo hello",
+			Env:     map[string]string{"FOO": "bar"},
+			Plugins: pipeline.Plugins{},
+			Secrets: nil, // No secrets
+		},
+		RepositoryURL: "https://github.com/example/repo",
+		OuterEnv:      map[string]string{"PIPELINE_VAR": "value"},
+	}
+
+	// Request fields without secrets - should succeed when step has no secrets
+	fields := []string{"command", "env", "plugins", "matrix", "repository_url"}
+	values, err := step.ValuesForFields(fields)
+	if err != nil {
+		t.Fatalf("ValuesForFields() unexpected error when step has no secrets and secrets field not requested: %v", err)
+	}
+
+	// Should have all the requested fields
+	if len(values) != 5 {
+		t.Errorf("ValuesForFields() returned %d fields, want 5", len(values))
+	}
+
+	// Should not have secrets field
+	if _, has := values["secrets"]; has {
+		t.Errorf("ValuesForFields() should not include 'secrets' field when step has no secrets")
+	}
```

Comment:
The focus here is on the secrets field, but perhaps we should diff the whole return value anyway?

```suggestion
	wantValues := map[string]any{
		"command": "echo hello",
		"env": map[string]string{"FOO": "bar"},
		"plugins": nil,
		"matrix": nil,
		"repository_url", "https://github.com/example/repo",
	}
	if diff := cmp.Diff(values, wantValues); diff != "" {
		t.Errorf(step.ValuesForFields(%v) diff (-got +want):\n%s", fields, diff)
	}
```